aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+public class S3OutputStream extends PositionOutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3OutputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final S3Client s3;
+  private final S3URI location;
+
+  private final OutputStream stream;
+  private final File stagingFile;
+  private long pos = 0;
+
+  private boolean closed = false;
+
+  public S3OutputStream(S3Client s3, S3URI location) throws IOException {
```

Comment:
Why throw `IOException` if the only place this constructor is called catches it and throws `UncheckedIOException`?

Reply:
This is just more of a convention, which I think makes sense to leave this way.  IO Streams typically throw the IOException (which is why this does because we're wrapping the buffered streams).  However, the OutputFile interface is really what's forcing the change in behavior because `create()` and `createAndOverwrite()` do not throw IOException and expect the an unchecked exception.